_posts/blog/2016-09-06-model-view-viewmodel.md

```diff
@@ -0,0 +1,127 @@
+
+* We should make sure that every action of the user goes through the ViewModel and that any possible logic of the View is moved in the ViewModel.
+
+We shared more details about these two topics in a blog post about <a href="https://upday.github.io/mvvm_rx_common_mistakes">common mistakes in MVVM + RxJava</a>.  
+
+### View
+
+The View is the actual user interface in the app. It can be an `Activity`, a `Fragment` or any custom Android `View`. For `Activities` and `Fragment`s, we are binding and unbinding from the event sources on `onResume()` and `onPause()`.
+
+{% highlight java %}
+    private final CompositeSubscription mSubscription = new CompositeSubscription();
+
+    @Override
+    public void onResume() {
+        super.onResume();
+        mSubscription.add(mViewModel.getSomeData()
+                         .observeOn(AndroidSchedulers.mainThread())
+                         .subscribe(this::updateView,
+                                    this::handleError));
+    }
+
+    @Override
+    public void onPause() {
+        mSubscription.clear();
+        super.onPause();
+    }
+{% endhighlight %}
+
+In case the MVVM View is a custom Android `View`, the binding is done in the constructor. To ensure that the subscription is not preserved, leading to possible memory leaks, the unbinding happens in `onDetachedFromWindow`.
+
+{% highlight java %}
+    private final CompositeSubscription mSubscription = new CompositeSubscription();
+
+    public MyView(Context context, MyViewModel viewModel) {
+        ...
+        mSubscription.add(mViewModel.getSomeData()
+                         .observeOn(AndroidSchedulers.mainThread())
+                         .subscribe(this::updateView,
+                                    this::handleError));
+    }
+
+    @Override
+    public void onDetachedFromWindow() {
+        mSubscription.clear();
+        super.onDetachedFromWindow();
+    }
+}
+{% endhighlight %}
+
+## Testability Of The Model-View-ViewModel Classes
+
+One of the main reasons why we love the Model-View-ViewModel pattern is its high degree of testability.
+
+### DataModel
+
+The inversion of control pattern, heavily applied in our code, and the lack of any Android classes, make the DataModel easy to unit test.
+
+### ViewModel
+
+The ViewModel exposes streams of data to whoever needs to consume it. We see the Views and the unit tests as two different types of data consumers.
+The ViewModel is completely separated from the UI or any Android classes, therefore straightforward to unit test.
+If the ViewModel needs access to Android classes, we create wrappers that we call `Provider`s. For example, for resources we created a `IResourceProvider`, that exposes methods like `String getString(@StringRes final int id)`. The implementation of the `IResourceProvider` will contain a reference to the `Context` but, the ViewModel will only get a reference to a `IResourceProvider` injected.
+
```

Comment:
maybe an example of a simple test? If I remember correctly you have some examples in your talk